File: packages/appkit-py/src/appkit_py/stream/sse_writer.py
"""SSE wire format writer matching the TypeScript SSEWriter.

Produces the exact format expected by the AppKit frontend:
    id: {uuid}
    event: {type}
    data: {json}
    (empty line)

Plus heartbeat comments: `: heartbeat\\n\\n`
"""

from __future__ import annotations

import json
import re
from typing import Any, Callable, Coroutine

from .types import BufferedEvent, SSEErrorCode, SSEWarningCode


def sanitize_event_type(event_type: str) -> str:
    """Sanitize SSE event type: remove newlines, cap at 100 chars."""
    sanitized = re.sub(r"[\r\n]", "", event_type)
    return sanitized[:100]


def format_event(event_id: str, event: dict[str, Any]) -> str:
    """Format a single SSE event as a string."""
    event_type = sanitize_event_type(str(event.get("type", "message")))
    event_data = json.dumps(event, separators=(",", ":"))
    return f"id: {event_id}\nevent: {event_type}\ndata: {event_data}\n\n"


def format_error(event_id: str, error: str, code: SSEErrorCode = SSEErrorCode.INTERNAL_ERROR) -> str:
    """Format an SSE error event."""
    data = json.dumps({"error": error, "code": code.value}, separators=(",", ":"))
    return f"id: {event_id}\nevent: error\ndata: {data}\n\n"


def format_buffered_event(event: BufferedEvent) -> str:
    """Format a buffered event for replay."""
    event_type = sanitize_event_type(event.type)
    return f"id: {event.id}\nevent: {event_type}\ndata: {event.data}\n\n"


def format_heartbeat() -> str:
    """Format an SSE heartbeat comment."""
    return ": heartbeat\n\n"


def format_buffer_overflow_warning(last_event_id: str) -> str:
    """Format a buffer overflow warning."""
    data = json.dumps({
        "warning": "Buffer overflow detected - some events were lost",
        "code": SSEWarningCode.BUFFER_OVERFLOW_RESTART.value,
        "lastEventId": last_event_id,
    }, separators=(",", ":"))
    return f"event: warning\ndata: {data}\n\n"


SSE_HEADERS = {
    "Content-Type": "text/event-stream",
    "Cache-Control": "no-cache",
    "Connection": "keep-alive",
    "Content-Encoding": "none",
    "X-Accel-Buffering": "no",
}

File: packages/appkit/src/stream/sse-writer.ts
import type { IAppResponse } from "shared";
import { streamDefaults } from "./defaults";
import {
  type BufferedEvent,
  type SSEError,
  SSEErrorCode,
  SSEWarningCode,
} from "./types";
import { StreamValidator } from "./validator";

export class SSEWriter {
  // setup SSE headers
  setupHeaders(res: IAppResponse): void {
    res.setHeader("Content-Type", "text/event-stream; charset=utf-8");
    res.setHeader("Cache-Control", "no-cache, no-transform");
    // X-Accel-Buffering: no — disables nginx-style proxy response buffering
    // for SSE (used by Cloudflare, AWS, GCP, and most corporate proxies).
    res.setHeader("X-Accel-Buffering", "no");
    // Intentionally NOT setting:
    //   - Connection: keep-alive   (HTTP/2 forbids it; Node manages keep-alive)
    //   - Content-Encoding: none   (invalid value; can trigger 502/RST in
    //                                strict intermediaries)
    res.flushHeaders?.();
    // Sentinel comment — a no-op SSE line that forces the response body
    // open immediately. Two purposes:
    //   1. Any buffering proxy must release the response headers + this
    //      first chunk to the client right away, so fetch() resolves with
    //      response.ok before the upstream generator yields.
    //   2. Prevents "Failed to fetch" symptoms where the browser gives up
    //      before the SQL query completes on cold-start warehouses.
    if (!res.writableEnded) {
      try {
        res.write(": ok\n\n");
      } catch {
        // ignore — handled by writeEvent's writableEnded check downstream
      }
    }
  }

  // write a single event to the response
  writeEvent(res: IAppResponse, eventId: string, event: any): void {
    if (res.writableEnded) return;

    const eventType = StreamValidator.sanitizeEventType(event.type);
    const eventData = JSON.stringify(event);

    res.write(`id: ${eventId}\n`);
    res.write(`event: ${eventType}\n`);
    res.write(`data: ${eventData}\n\n`);
  }
  writeError(
    res: IAppResponse,
    eventId: string,
    error: string,
    code: SSEErrorCode = SSEErrorCode.INTERNAL_ERROR,
  ): void {
    if (res.writableEnded) return;

    const errorData: SSEError = {
      error,
      code,
    };

    res.write(`id: ${eventId}\n`);
    res.write(`event: error\n`);
    res.write(`data: ${JSON.stringify(errorData)}\n\n`);
  }

  // write a buffered event for replay
  writeBufferedEvent(res: IAppResponse, event: BufferedEvent): void {
    if (res.writableEnded) return;

    res.write(`id: ${event.id}\n`);
    res.write(`event: ${event.type}\n`);
    res.write(`data: ${event.data}\n\n`);
  }

  // write a buffer overflow warning
  writeBufferOverflowWarning(res: IAppResponse, lastEventId: string): void {
    if (res.writableEnded) return;

    try {
      res.write(`event: warning\n`);
      res.write(
        `data: ${JSON.stringify({
          warning: "Buffer overflow detected - some events were lost",
          code: SSEWarningCode.BUFFER_OVERFLOW_RESTART,
          lastEventId,
        })}\n\n`,
      );
    } catch (_error) {
      // ignore write errors - client will ignore this event
    }
  }

  // start the heartbeat interval
  startHeartbeat(
    res: IAppResponse,
    signal: AbortSignal,
    interval?: number,
  ): NodeJS.Timeout {
    const heartbeatInterval = interval ?? streamDefaults.heartbeatInterval;

    return setInterval(() => {
      if (!signal.aborted && !res.writableEnded) {
        try {
          res.write(`: heartbeat\n\n`);
        } catch (_error) {
          // ignore write errors - client will ignore this event
        }
      }
    }, heartbeatInterval);
  }
}
